entity-framework/core/testing/choosing-a-testing-strategy.md

```diff
@@ -0,0 +1,112 @@
+---
+title: Choosing a testing strategy - EF Core
+description: Different approaches to testing applications that use Entity Framework Core
+author: roji
+ms.date: 11/07/2021
+uid: core/testing/choosing-a-testing-strategy
+---
+# Choosing a testing strategy
+
+As discussed in the [Overview](xref:core/testing/index), we'll refer to tests which involve your production database system as *integration tests*, and tests which employ [test doubles](https://martinfowler.com/bliki/TestDouble.html) as *unit tests*.
+
+Integration tests frequently involve the following difficulties:
+
+1. In many cases, it's simply not possible or practical to test against the actual external dependency. For example, your application may interact with some service that cannot be easily tested against (because of rate limiting, or the lack of a testing environment).
+2. Even when it's possible to involve the real dependency, this may be exceedingly slow: running a large amount of tests against a cloud service may cause tests to take too long. Testing should be part of the developer's everyday workflow, so it's important that tests run quickly.
+3. Executing tests against an external resource may involve isolation issues, where tests interfere with one another. For example, multiple tests running in parallel against a database may modify data and cause each other to fail in various ways. Since unit tests run in-memory and don't use shared resources, they're isolated from each other and don't cause interference.
+
+However, unit tests do not guarantee that your program works correctly as a whole; while each component may work well in isolation, various issues may arise when integrating them together. For example, a component may execute some query against a database, expecting certain behavior, while the actual database system behaves differently. Such issues are only uncovered in integration testing, making them an important part of any testing strategy. In addition, since unit tests interact with internal components, they must be reworked as the program is refactored and internal APIs change; integration tests, in contrast, are more stable as they test public, external API contracts.
+
+## Integration testing
+
+Because of the above difficulties with integration tests, developers are frequently urged to concentrate on unit tests first, and have a robust unit test suite which they can run frequently on their machines; integration tests, in contrast, are supposed to be executed much less frequently, and in many cases also provide far less coverage. We recommend giving more thought to integration tests, and suggest that databases may actually be far less affected by the above problems as much as people tend to think:
```

Comment:
```suggestion
Because of the above difficulties with integration tests, developers are frequently urged to concentrate on unit tests first, and have a robust unit test suite which they can run frequently on their machines; integration tests, in contrast, are supposed to be executed much less frequently, and in many cases also provide far less coverage. We recommend giving more thought to integration tests, and suggest that databases may actually be far less affected by the above problems than people tend to think:
```